sdks/python/apache_beam/io/textio.py

```diff
@@ -527,6 +533,61 @@ def expand(self, pvalue):
     return pvalue.pipeline | Read(self._source)
 
 
+class ReadFromTextWithFilename(PTransform):
```

Comment:
Perhaps it would make more sense to allow returning filenames be an option for all file-based sources, rather than a new transform? At the very least, much of the redundancy here could be removed.

Reply:
This is definitely an option. I implemented it in this way because I found strage have a class that change the returned data type format based on an optional parameter. 

In this way, it is more clear (and documented) that if I want to get the source filename, I will need to change the class **and** the way I handle elements in the following steps.

Reply:
Yes, I agree with that sentiment. The disadvantage I see here is that there will have to be an active effort to keep the (non-trivial) set of arguments in sync between the two from now on. 

Another option would be to have a _ReadFromTextBase with two subclasses that differ only in a class attribute set to either _TextSource or _FilenameTextSource.

Reply:
I like the _ReadFromTextBase approach, I'm working to update the code accordingly.